Add cost inference to ConvGradient and WeightedSum operators

mdschatz opened the pull request:
Summary: As title

Differential Revision: D9436387

jspark1105 commented on caffe2/operators/conv_gradient_op.cc:
```diff
@@ -4,8 +4,85 @@
+std::vector<TensorShape> TensorInferenceForConvGradient(
+    const OperatorDef& def,
+    const std::vector<TensorShape>& in) {
+  CAFFE_ENFORCE(in.size() >= 2, "ConvGradient expects at least 2 inputs");
```
This should be 3 same as CostInferenceForConvGradient?
``
  CAFFE_ENFORCE_EQ(in.size(), 3, "ConvGradient expects 3 inputs");
``

jspark1105 commented on caffe2/operators/conv_gradient_op.cc:
```diff
@@ -4,8 +4,85 @@
+  auto i = 0;
+  vector<TensorShape> out(n_outputs);
+
+  out[i++] = in[1];
```
Maybe it's just me but i++ for indexing confuses me. Perhaps out[i] = in[1]; ++i; ?

``
  // FILTER_GRAD has the same shape as FILTER
  out[i] = in[1];
  ++i;
``

jspark1105 commented on caffe2/operators/conv_gradient_op.cc:
```diff
@@ -4,8 +4,85 @@
+  }
+
+  if (n_outputs == 3 || (no_bias && n_outputs == 2)) {
+    out[i++] = in[0];
```
Add comment:
// INPUT_GRAD has the same shape as INPUT.

jspark1105 commented on caffe2/operators/conv_gradient_op.cc:
```diff
@@ -4,8 +4,85 @@
+OpSchema::Cost CostInferenceForConvGradient(
+    const OperatorDef& def,
+    const vector<TensorShape>& inputs) {
+  CAFFE_ENFORCE_EQ(inputs.size(), 3, "ConvGradient requires at least 3 inputs");
```
``
CAFFE_ENFORCE_EQ(inputs.size(), 3, "ConvGradient expects 3 inputs");
``

jspark1105 commented on caffe2/operators/conv_gradient_op.cc:
```diff
@@ -4,8 +4,85 @@
+  const auto& dY = inputs[2];
+  const auto N = X.dims(0);
+  const auto M = filter.dims(0);
+  const auto C = (order == StorageOrder::NCHW ? X.dims(1) : X.dims(3));
```
@BIT-silence is working on 3D conv in NHWC so let's make it general.
``
  const auto C = (order == StorageOrder::NCHW ? X.dims(1) : X.dims(X.dims_size() - 1));
``

jspark1105 commented on caffe2/operators/conv_gradient_op.cc:
```diff
@@ -4,8 +4,85 @@
+  const auto C = (order == StorageOrder::NCHW ? X.dims(1) : X.dims(3));
+  const auto output_image_size =
+      (order == StorageOrder::NCHW ? nElemFromDim(dY, 2)
+                                   : dY.dims(1) * dY.dims(2));
```
Perhaps we should have a function like nElemBetweenDim like size_between_dim function for Tensor and use ``eElemBetweenDim(dY, 1, dY.dims_size() - 1)`` ?

jspark1105 commented on caffe2/operators/conv_gradient_op.cc:
```diff
@@ -4,8 +4,85 @@
+                                   : dY.dims(1) * dY.dims(2));
+  auto kernel_elem =
+      (order == StorageOrder::NCHW ? nElemFromDim(filter, 2)
+                                   : filter.dims(1) * filter.dims(2));
```
Ditto.

jspark1105 commented on caffe2/operators/conv_gradient_op.cc:
```diff
@@ -4,8 +4,85 @@
+  struct OpSchema::Cost c;
+  c.flops = N * 2 * M * kernel_elem * C * output_image_size;
+  if (!no_bias) {
+    c.flops += N * (2 * M * output_image_size);
```
There shouldn't be a factor of 2 because all elements in bias_multiplier are just 1 so ideally we don't need multiplications. Also this is to be consistent with FCGradient.

jspark1105 commented on caffe2/operators/conv_gradient_op.cc:
```diff
@@ -4,8 +4,85 @@
+    c.flops += N * 2 * M * kernel_elem * C * output_image_size;
+  }
+
+  c.bytes_read = (nElemFromDim(X) + nElemFromDim(filter) + nElemFromDim(dY)) *
```
Not for this diff, but I see bytes_read is missing in FCGradient cost function.

jspark1105 commented on caffe2/operators/conv_gradient_op.cc:
```diff
@@ -4,8 +4,85 @@
+  for (auto i = 0; i < n_outputs; i++) {
+    c.bytes_written += nElemFromDim(outputs[i]) * sizeof(float);
+  }
+  c.params_bytes = nElemFromDim(filter) * sizeof(float);
```
Not for this diff, but I see params_bytes comment in operator_schema.h is still outdated. We should update from

``
uint64_t params_bytes{0}; // Memory footprint of parameters
``
to
``
uint64_t params_bytes{0}; // Memory read for parameters.
``

mdschatz replied:
I see you've addressed this :).

jspark1105 commented on caffe2/operators/utility_ops.cc:
```diff
@@ -14,6 +14,31 @@ bool WeightedSumGradientOp<CPUContext>::RunOnDevice() {
+  const auto& X0 = in[0];
+  const auto& nElem = nElemFromDim(X0);
+  const auto& nInputs = in.size();
+  c.flops = (nInputs - 1) * nElem;
```
We also have multiplications.
``
c.flops = (2 * nInputs - 1) * nElem;
``

mdschatz replied:
The count is correct as is.  The formulation should really be `(2 * nInputs / 2 - 1) * nElem`.  The `nInputs / 2` accounts for the number of pairs.

Commit:
Add cost inference to ConvGradient and WeightedSum operators (#10744)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/10744

As title

Reviewed By: jspark1105

Differential Revision: D9436387

fbshipit-source-id: 58d500b163aa4bfd181e013790f2294cb9e26026
